design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+0. Modify spilling in new TLogServer
+0. Modify peeking in new TLogServer
+0. Modify popping in new TLogServer
+0. Spill txsTag specially
+
+### Spilling
+
+In spill-by-reference, spilling is now the act of persisting the index of 
+
+### Peeking
+
+A `TLogPeekRequest` contains a `Tag` and a `Version`, and is a request for all
+commits with the specified tag with a commit version greater than or equal to
+the given version.  The goal is to return a 150KB block of mutations.
+
+**************************************************************************
+*                                                                        *
+* +---------+  Tag    +---------+           Tag    +--------+            *
+* |  Peek   |-------->| Spilled | ...------------->| Memory |            *
+* | Request | Version |  Index  |          Version | Index  |            *
+* +---------+         +---------+                  +--------+            *
+*                        |                                 |             *
+*      +-----------------+-----------------+               |             *
+*     / \  Start=100   _/ \_  Start=500    +  Start=900    +  Ptr=0xF00  *
+*    /   \ Length=50  /     \ Length=70   / \ Length=30   / \ Length=30  *
+*  +------------------------------------------------+------------------+ *
+*  |                   Disk Queue                   |  Also In Memory  | *
+*  +------------------------------------------------+------------------+ *
+*                                                                        *
+**************************************************************************
+
+Spill-by-value and memory storage engine only ever read from DiskQueue when
+recovering, and read the entire file linearly.  Therefore, `IDiskQueue` had no
+API for random reads to the DiskQueue.  That ability is now required for
+peeking, and thus, `IDiskQueue`'s API has been enhanced correspondingly:
+
+``` CPP
+enum class CheckHashes { NO, YES };
+
+class IDiskQueue {
+    // ...
+    Future<Standalone<StringRef>> read(location start, location end, CheckHashes ch);
+    // ...
+};
+```
+
+Internally, the DiskQueue adds page headers every 4K, which are stripped out
+from the returned data.  Therefore, the length of the result will not be the
+same as `end-start`, intentionally.  For this reason, the API is `(start, end)`
+and not `(start, length)`.
+
+Spilled data, when using spill-by-value, was resistent to bitrot via data being
+checksummed interally within SQLite's B-tree.  Now that reads can be done
+directly, the responsibility for verifing data integrity falls upon the
+DiskQueue.  `CheckHashes::YES` will cause the DiskQueue to use the checksum in
+each DiskQueue page to verify data integrity.  If an externally maintained
+checksums exists to verify the returned data, then `CheckHashes::NO` can be
+used to elide the checksumming.
+
+### Popping
+
+As storage servers persist data, they send `pop(tag, version)` requests to the
```

Comment:
Ah ok I was wondering how the cleanup process worked. I'd probably move this up higher towards the beginning of the doc too because its the first question I immediately had about how the disk queue worked and helps explain the interaction between the `spill loop` and the `diskqueue`. I.E the disk queue is normally popped by the storage nodes themselves, but in the case that a storage node is down, the spill loop needs to issue the pop requests on that storage nodes behalf (once the data for that node has been safely tucked away in the spilling BTree) otherwise the disk queue will grow infinitely